File: common/src/main/java/io/xol/chunkstories/entity/EntityDefinitionsStore.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.entity;

import java.io.BufferedReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.xol.chunkstories.api.content.Asset;
import io.xol.chunkstories.api.content.Content;
import io.xol.chunkstories.api.content.Content.EntityDefinitions;
import io.xol.chunkstories.api.entity.EntityDefinition;
import io.xol.chunkstories.api.exceptions.content.IllegalEntityDeclarationException;
import io.xol.chunkstories.content.GameContentStore;

public class EntityDefinitionsStore implements EntityDefinitions {
	private final Content content;

	private Map<String, EntityDefinition> EntityDefinitionsByName = new HashMap<String, EntityDefinition>();

	private static final Logger logger = LoggerFactory.getLogger("content.entities");

	public Logger logger() {
		return logger;
	}

	public EntityDefinitionsStore(GameContentStore content) {
		this.content = content;
	}

	public void reload() {
		// EntityDefinitionsById.clear();
		EntityDefinitionsByName.clear();

		Iterator<Asset> i = content.modsManager().getAllAssetsByExtension("entities");
		while (i.hasNext()) {
			Asset f = i.next();
			readEntitiesDefinitions(f);
		}

		// this.entityComponents.reload();
	}

	private void readEntitiesDefinitions(Asset f) {
		if (f == null)
			return;

		logger().debug("Reading entities definitions in : " + f);
		try {
			BufferedReader reader = new BufferedReader(f.reader());
			String line = "";
			while ((line = reader.readLine()) != null) {
				line = line.replace("\t", "");
				if (line.startsWith("#")) {
					// It's a comment, ignore.
				} else {
					if (line.startsWith("entity ")) {
						String[] split = line.split(" ");
						String name = split[1];
						// short id = Short.parseShort(split[2]);

						try {
							EntityDefinitionImplementation entityType = new EntityDefinitionImplementation(this, name,
									reader);

							// this.EntityDefinitionsById.put(entityType.getId(), entityType);
							this.EntityDefinitionsByName.put(entityType.getName(), entityType);
						} catch (IllegalEntityDeclarationException e) {
							e.printStackTrace();
						}
					}
				}
			}
			reader.close();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	@Override
	public EntityDefinition getEntityDefinition(String TraitName) {
		return EntityDefinitionsByName.get(TraitName);
	}

	@Override
	public Iterator<EntityDefinition> all() {
		return this.EntityDefinitionsByName.values().iterator();
	}

	@Override
	public Content parent() {
		return content;
	}
}

File: common/src/main/java/io/xol/chunkstories/entity/EntityDefinitionsStore.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.entity

import java.util.HashMap

import org.slf4j.Logger
import org.slf4j.LoggerFactory

import DefinitionsLexer
import DefinitionsParser
import io.xol.chunkstories.api.content.Asset
import io.xol.chunkstories.api.content.Content
import io.xol.chunkstories.api.content.Content.EntityDefinitions
import io.xol.chunkstories.api.entity.EntityDefinition
import io.xol.chunkstories.content.GameContentStore
import io.xol.chunkstories.util.format.toMap
import org.antlr.v4.runtime.ANTLRInputStream
import org.antlr.v4.runtime.CommonTokenStream

class EntityDefinitionsStore(content: GameContentStore) : EntityDefinitions {
    private val content: Content

    private val entityDefinitions = HashMap<String, EntityDefinition>()

    override fun logger(): Logger {
        return logger
    }

    init {
        this.content = content
    }

    fun reload() {
        entityDefinitions.clear()

        fun readEntitiesDefinitions(a: Asset) {
            logger().debug("Reading entities definitions in : $a")

            val text = a.reader().use { it.readText() }
            val parser = DefinitionsParser(CommonTokenStream(DefinitionsLexer(ANTLRInputStream(text))))

            for(definition in parser.entitiesDefinitions().entitiesDefinition()) {
                val name = definition.Name().text
                val properties = definition.properties().toMap()

                val entityDefinition = EntityDefinition(this, name, properties)
                entityDefinitions.put(name, entityDefinition)

                logger().debug("Loaded entity definition $entityDefinition")
            }
        }

        val i = content.modsManager().getAllAssetsByExtension("entities")
        while (i.hasNext()) {
            val f = i.next()
            readEntitiesDefinitions(f)
        }
    }

    override fun getEntityDefinition(TraitName: String): EntityDefinition? {
        return entityDefinitions[TraitName]
    }

    override fun all(): Iterator<EntityDefinition> {
        return this.entityDefinitions.values.iterator()
    }

    override fun parent(): Content {
        return content
    }

    companion object {

        private val logger = LoggerFactory.getLogger("content.entities")
    }
}
